backends/dpdk/dpdkProgram.cpp

```diff
@@ -149,6 +170,10 @@ cstring ConvertToDpdkParser::append_parser_name(const IR::P4Parser *p, cstring l
     return p->name + "_" + label;
 }
 
+void ConvertToDpdkParser::add_metadata_field(IR::Declaration_Variable *d) {
+    metadataStruct->fields.push_back(new IR::StructField(IR::ID(d->name.name), d->type));
```

Comment:
what happens if a local variable has a name identical to one of the existing fields?

Reply:
I assumed getting a unique name using refMap would take care of any name conflicts. Please correct me if I am wrong and please also suggest how we should handle it in this particular case(creating a new temporary variable and inserting it into an existing Structure).

Also, as per my understanding of the dpdk backend code, it is not possible to have an identical name in the metadata struct field from elsewhere. The metadata struct fields from user program are prefixed with "local_metadata_" and the locals within different parser/control blocks are prefixed with the block name. Ex: a variable named "tmpMask" in "Ingress" block is inserted in metadata struct as "Ingress_tmpMask_".

I am thinking if these new temporary variables for Mask operations should also be prefixed with something similar?

Reply:
Looking at the generated code can you reverse engineer an input program that would produce the wrong results because it generates the same name twice? If you can, you should add that program as a test case and fix the code to make sure it uses everywhere unique identifiers. In this particular line you are not generating a fresh name, you are using the variable name directly.

Reply:
A fresh name is created right before calling this function and this function is called only from one place. 


                         auto tmpDecl = new IR::Declaration_Variable(IR::ID(
                                                 refmap->newName("tmpMask")), switch_var->type);

                            .......
                         add_metadata_field(tmpDecl);

I tried a few more modifications to my test program to use the variable with same name in different parser and control blocks. There seems to be no conflict.

But I agree, to use add_metadata_field to be called from other places, we will have to make sure that the caller creates a fresh name. To fix this, I will modify the function to create a new field name, insert it to metadata and return the same declaration to the caller. The caller can then use the returned declaration to create a pathexpression.